File: OC/20190606/Configuration/UnitClass/Sandbox/XXZSandBox.m
//
//  XXZSandBox.m
//  Author From Xavier Zachary
//
//  Created by Zachary on 14/10/29.
//  Copyright © 2015年 com.xxzing.future. All rights reserved.
//

#import "XXZSandBox.h"

@implementation XXZSandBox

//创建的文件或文件夹 存放在Documents文件夹下
//获取文件或文件夹路径, 若是文件则含文件类型, 若是文件夹则只有文件夹名
+ (NSString *)documentsWithFileName:(NSString *)fileName folderName:(NSString *)folderName {
    //获取根目录
    NSString *home = NSHomeDirectory();
    NSString *appendName = nil;
    
    //拼接
    if (folderName.length && fileName.length) {//创建文件夹下的文件
        appendName = [NSString stringWithFormat:@"Documents/%@/%@", folderName, fileName];
    }
    else if (fileName.length) { //创建文件
        appendName = [NSString stringWithFormat:@"Documents/%@",fileName];
    }
    else if (folderName.length) {//创建文件夹
        appendName = [NSString stringWithFormat:@"Documents/%@", folderName];
    }
    else { //无意义
        return nil;
    }
    
    //拼接完整路径
    NSString *filePath = [home stringByAppendingPathComponent:appendName];
    XXZLog(@"filePath = %@", filePath);
    
    return filePath;
}

//创建文件或文件夹, 若是文件则含文件类型, 若是文件夹则只有文件夹名
+ (void)createFileWithFileName:(NSString *)fileName folderName:(NSString *)folderName {
    NSInteger createType = -1;
    if (!fileName.length && !folderName.length) {//无意义
        XXZLog(@"文件和文件夹没有名称..");
        return;
    }
    
    NSString *completePath = [self documentsWithFileName:fileName folderName:folderName];
    NSFileManager *fm = [NSFileManager defaultManager];
    
    if (folderName.length && !fileName.length) { //文件夹路径
        createType = 0;
    }
    else if (fileName.length && !folderName.length) { //文件路径
        createType = 1;
    }
    else if (folderName.length && fileName.length) { //文件夹下的文件路径
        createType = 2;
    }
    else { //无意义
        return;
    }
    
    
    if (![fm fileExistsAtPath:completePath]) {
        if (createType == 0) { //创建文件夹
            if (![fm fileExistsAtPath:completePath]) { //若不存在, 创建
                BOOL flag = [fm createDirectoryAtPath:completePath withIntermediateDirectories:YES attributes:nil error:nil];
                if (!flag) {
                    XXZLog(@"文件夹创建失败");
                }
            }
        }
        else if (createType == 1) { //创建文件
            if (![fm fileExistsAtPath:completePath]) { //若不存在, 创建
                BOOL flag = [fm createFileAtPath:completePath contents:nil attributes:nil];
                if (!flag) {
                    XXZLog(@"文件创建失败");
                }
            }
        }
        else if (createType ==2) { //创建文件夹下的文件
            NSString *folderPath = [self documentsWithFileName:nil folderName:folderName];
            if (![fm fileExistsAtPath:folderName]) {//创建文件夹
                BOOL flag = [fm createDirectoryAtPath:folderPath withIntermediateDirectories:YES attributes:nil error:nil];
                if (!flag) {
                    XXZLog(@"文件夹创建失败");
                }
                else {
                    NSString *filePath = [self documentsWithFileName:fileName folderName:folderPath];
                    if (![fm fileExistsAtPath:filePath]) { //创建该文件夹下的文件
                        BOOL flag = [fm createFileAtPath:completePath contents:nil attributes:nil];
                        if (!flag) {
                            XXZLog(@"文件夹下的文件创建失败");
                        }
                    }
                }
                
            }
        }
        else {//无意义
            return;
        }
    }
    else{
        XXZLog(@"file have existed : \n%@",completePath);
    }
}

//移除文件夹及文件夹内的文件：
+ (BOOL)removeFolderWithFolderPath:(NSString *)folderPath {
    NSFileManager *fm = [NSFileManager defaultManager];
    
    if ([fm fileExistsAtPath:folderPath]) { //若存在,移除
        BOOL success = [fm removeItemAtPath:folderPath error:nil];
        if (success) {
            return YES;
        }
        else {
            XXZLog(@"移除失败..");
        }
    }
    else {
        XXZLog(@"路径不存在..");
    }
    
    return NO;
}

//移除文件
+ (BOOL)removeFileWithFilePath:(NSString *)filePath {
    NSFileManager *fm = [NSFileManager defaultManager];
    
    if ([fm fileExistsAtPath:filePath]) { //若存在,移除
        BOOL success = [fm removeItemAtPath:filePath error:nil];
        if (success) {
            return YES;
        }
        else {
            XXZLog(@"移除失败..");
        }
    }
    else {
        XXZLog(@"路径不存在..");
    }
    
    return NO;
}

//删除文件夹指定类型的文件
+ (BOOL)removeFileWithFolderPath:(NSString *)folderPath extension:(NSString *)extension {
    if (!extension.length) {
        XXZLog(@"没有输入扩展名..");
        return NO;
    }
    
    NSFileManager *fm = [NSFileManager defaultManager];
    if (![fm fileExistsAtPath:folderPath]) {
        XXZLog(@"文件夹路径不存在!");
        return NO;
    }
    
    //获取文件夹下的所有文件 => test.plist
    NSArray *contents = [fm contentsOfDirectoryAtPath:folderPath error:NULL];
    if (!contents.count) { //没有, 直接返回
        return YES;
    }
    
    NSEnumerator *e = [contents objectEnumerator];
    NSString *fileName;
    
    while ((fileName = [e nextObject])) {
        NSString *pathExtension = [fileName pathExtension];//取出扩展名
        
        if ([pathExtension isEqualToString:extension]) {//删除符合条件的文件
            BOOL flag = [fm removeItemAtPath:[folderPath stringByAppendingPathComponent:fileName] error:NULL];
            
            if (!flag) {//删除失败,直接返回
                XXZLog(@"移除(%@)失败..", fileName);
                return NO;
            }
        }
    }
    
    return YES;
}

//拷贝
+ (BOOL)copyItemAtPath:(NSString *)srcPath toPath:(NSString *)dstPath {
    NSFileManager *fm = [NSFileManager defaultManager];
    
    if (![fm fileExistsAtPath:srcPath] || ![fm fileExistsAtPath:dstPath]) {
        XXZLog(@"路径错误..");
        return NO; //是否存在资源路径 //是否存在目的地路径
    }
    
    //拷贝
    NSString *finalPath = [dstPath stringByAppendingPathComponent:[srcPath lastPathComponent]];
    BOOL success = [fm copyItemAtPath:srcPath toPath:finalPath error:nil];
    if (!success) {
        XXZLog(@"拷贝失败..");
        return NO;
    }
    
    return YES;
}

//移动
+ (BOOL)moveItemAtPath:(NSString *)srcPath toPath:(NSString *)dstPath {
    NSFileManager *fm = [NSFileManager defaultManager];
    
    if (![fm fileExistsAtPath:srcPath] || ![fm fileExistsAtPath:dstPath]) {
        XXZLog(@"路径错误..");
        return NO; //是否存在资源路径 //是否存在目的地路径
    }
    
    //移动
    NSString *finalPath = [dstPath stringByAppendingPathComponent:[srcPath lastPathComponent]];
    BOOL success = [fm moveItemAtPath:srcPath toPath:finalPath error:nil];
    if (!success) {
        XXZLog(@"移动失败..");
        return NO;
    }
    
    return YES;
}

@end

File: Swift/20190827/Configuration/UnitClass/Sandbox/XXZSandbox.swift
//
//  XXZSandbox.swift
//  Author From Xavier Zachary
//
//  Created by Zachary on 2017/7/31.
//  Copyright © 2017年 Zachary. All rights reserved.
//

import UIKit

class XXZSandbox: NSObject {
    private override init() {}
    
    static func moveItemAtPath(_ src:String?, _ dst:String) { //移动
        if let tempSrc = src {
            let fm = FileManager.default
            if fm.fileExists(atPath: dst) {
                XXZLog("dst file path have exist")
                return;
            }
            
            try! fm.moveItem(atPath: tempSrc, toPath: dst)
        }
        else {
            XXZLog("src file path have no exist")
        }
    }
    
    static func copyItemAtPath(_ src:String?, _ dst:String) { //拷贝
        if let tempSrc = src {
            let fm = FileManager.default
            
            if fm.fileExists(atPath: dst) {
                XXZLog("dst file path have exist")
                return;
            }
            
            try! fm.copyItem(atPath: tempSrc, toPath: dst)
        }
        else {
            XXZLog("src file path have no exist")
        }
    }
    
    static func removeFileWithFolder(_ extensionName:String, _ folderName:String) { //移除指定扩展名的所有文件
        if extensionName.isBlank() {
            XXZLog("no file extension name")
            return;
        }
        
        var completePath = ""
        if folderName.isBlank() { //没有文件夹
            completePath = NSHomeDirectory().appending("/Documents")
        }
        else { //有文件夹
            completePath = NSHomeDirectory().appending("/Documents/\(folderName)")
        }
        
        let fm = FileManager.default
        
        //获取文件夹下的所有文件
        let fileArr = try! fm.contentsOfDirectory(atPath: completePath)
        if fileArr.count <= 0 { //没有文件
            return;
        }
        
        for i in 0..<fileArr.count {
            let filePath = fileArr[i]
            if filePath .hasSuffix(extensionName) {
                try! fm.removeItem(atPath: completePath.appending("/\(filePath)"))
            }
        }
    }
    
    static func removeFileWithPath(_ fileName:String, _ folderName:String = "") { //移除指定文件夹内的文件
        let fm = FileManager.default
        
        var completePath = ""
        if folderName.isBlank() { //没有文件夹
            completePath = NSHomeDirectory().appending("/Documents/\(fileName)")
        }
        else { //有文件夹
            completePath = NSHomeDirectory().appending("/Documents/\(folderName)/\(fileName)")
        }
        
        if fm.fileExists(atPath: completePath) { //若存在, 移除
            try! fm.removeItem(atPath: completePath)
        }
        else {
            XXZLog("file is not exist")
        }
    }
    
    static func removeFolderWithPath(_ folderName:String) { //移除文件夹及文件夹内的文件
        let fm = FileManager.default
        
        let completePath = NSHomeDirectory().appending("/Documents/\(folderName)")
        if fm.fileExists(atPath: completePath) { //若存在, 移除
            try! fm.removeItem(atPath: completePath)
        }
        else {
            XXZLog("folder is not exist")
        }
    }
    
    static func createWithPath(_ fileName:String, _ folderName:String = "") { //根据文件名及文件夹名 创建文件及文件夹
        var createType = 0
        if !fileName.isBlank() && !folderName.isBlank() { //创建文件及文件夹
            createType = 1
        }
        else if !fileName.isBlank() { //只创建文件
            createType = 2
        }
        else if !folderName.isBlank() { //只创建文件夹
            createType = 3
        }
        else {
            XXZLog("no meaning")
        }
        
        let completePath = self.pathWithName(fileName, folderName)
        let fm = FileManager.default
        
        if createType == 1 {
            //创建文件夹
            let folderPath = self.pathWithName("", folderName)
            if !fm.fileExists(atPath: folderPath) {
                try! fm.createDirectory(atPath: folderPath, withIntermediateDirectories: true, attributes: nil)
            }
            else {
                XXZLog("folder have exist")
            }
            
            //创建文件
            if !fm.fileExists(atPath: completePath) {
                let flag = fm.createFile(atPath: completePath, contents: nil, attributes: nil)
                if !flag {
                    XXZLog("create file failed")
                }
            }
            else {
                XXZLog("file have exist")
            }
        }
        else if createType == 2 { //只创建文件
            if !fm.fileExists(atPath: completePath) {
                let flag = fm.createFile(atPath: completePath, contents: nil, attributes: nil)
                if !flag {
                    XXZLog("create file failed")
                }
            }
            else {
                XXZLog("file have exist")
            }
        }
        else if createType == 3 { //只创建文件夹
            if !fm.fileExists(atPath: completePath) {
                try! fm.createDirectory(atPath: completePath, withIntermediateDirectories: true, attributes: nil)
            }
            else {
                XXZLog("folder have exist")
            }
        }
        else {
            XXZLog("no meaning")
        }
    }
    
    static func pathWithName(_ file:String, _ folder:String = "") -> String { //根据文件名及文件夹名 获取文件路径
        let home = NSHomeDirectory()
        
        var appendName = ""
        if !file.isBlank() && !folder.isBlank() { //有文件名和文件夹名
            appendName = String.init(format: "Documents/%@/%@", folder, file)
        }
        else if !file.isBlank() { //只有文件名
            appendName = String.init(format: "Documents/%@", file)
        }
        else if !folder.isBlank() { //只有文件夹名
            appendName = String.init(format: "Documents/%@", folder)
        }
        else {
            XXZLog("no meaning")
            return ""
        }
        
        //拼接完整路径
        let path = home.appending("/\(appendName)")
        XXZLog("\(path)")
        
        return path
    }

}
